File: core/src/no/elg/infiniteBootleg/input/KeyboardControls.java
package no.elg.infiniteBootleg.input;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.Input;
import com.badlogic.gdx.math.Vector2;
import com.badlogic.gdx.physics.box2d.Body;
import no.elg.infiniteBootleg.Main;
import no.elg.infiniteBootleg.world.Block;
import no.elg.infiniteBootleg.world.Material;
import no.elg.infiniteBootleg.world.World;
import no.elg.infiniteBootleg.world.render.WorldRender;
import no.elg.infiniteBootleg.world.subgrid.Entity;
import no.elg.infiniteBootleg.world.subgrid.LivingEntity;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import static com.badlogic.gdx.Input.Keys.*;
import static no.elg.infiniteBootleg.world.render.WorldRender.BOX2D_LOCK;

/**
 * Control scheme where the user moves the player around with a keyboard
 *
 * @author Elg
 */
public class KeyboardControls extends AbstractEntityControls {

    public static final float JUMP_VERTICAL_IMPULSE = .5f;
    public static final float HORIZONTAL_IMPULSE = .15f;
    public static final int EDIT_TICK_DELAY = 1; //delay in ticks between allowing to place/break blocks

    private Material selected;
    //if objects can be placed on non-air blocks
    public boolean replacePlacement;
    private float breakBrushSize = 2f;
    private float placeBrushSize = 1f;
    private long lastEditTick;

    public KeyboardControls(@NotNull WorldRender worldRender, @NotNull LivingEntity entity) {
        super(worldRender, entity);
        selected = Material.STONE;
    }

    @Override
    public void update() {
        if (Main.inst().getConsole().isVisible()) {
            return;
        }

        boolean update = false;
        int blockX = Main.inst().getMouseBlockX();
        int blockY = Main.inst().getMouseBlockY();

        float rawX = Main.inst().getMouseX();
        float rawY = Main.inst().getMouseY();
        World world = getWorldRender().getWorld();

        if (Gdx.input.isButtonPressed(Input.Buttons.LEFT)) {

            if (breakBrushSize <= 1) {
                world.remove(blockX, blockY, true);
            }
            else {
                for (Block block : world.getBlocksWithin(rawX, rawY, breakBrushSize, true)) {
                    world.remove(block.getWorldX(), block.getWorldY(), true);
                }
            }
            lastEditTick = world.getTick();
            update = true;
        }
        else if (Gdx.input.isButtonPressed(Input.Buttons.RIGHT) || Gdx.input.isKeyJustPressed(Q)) {

            if (placeBrushSize <= 1) {
                update = selected.create(world, blockX, blockY);
            }
            else {
                for (Block block : world.getBlocksWithin(rawX, rawY, placeBrushSize, false)) {
                    update |= selected.create(world, block.getWorldX(), block.getWorldY());
                }
            }
            if (update) {
                lastEditTick = world.getTick();
            }
        }

        Entity entity = getControlled();

        if (Gdx.input.isKeyJustPressed(Input.Keys.T)) {
            //teleport the player to the (last) location of the mouse
            entity.teleport(Main.inst().getMouseX(), Main.inst().getMouseY(), true);
            WorldInputHandler input = world.getInput();
            if (input != null) {
                input.setFollowing(entity);
                input.setLockedOn(true);
            }
        }
        else {
            boolean shift = Gdx.input.isKeyPressed(Input.Keys.SHIFT_LEFT);

            if (entity.isFlying()) {

                float modifier = shift ? 2f : 1f;

                if (Gdx.input.isKeyPressed(Input.Keys.W)) {
                    setVel(0, HORIZONTAL_IMPULSE * modifier);
                }
                if (Gdx.input.isKeyPressed(Input.Keys.S)) {
                    setVel(0, -HORIZONTAL_IMPULSE * modifier);
                }


                if (Gdx.input.isKeyPressed(Input.Keys.A)) {
                    setVel(-HORIZONTAL_IMPULSE * modifier, 0);
                }
                if (Gdx.input.isKeyPressed(Input.Keys.D)) {
                    setVel(HORIZONTAL_IMPULSE * modifier, 0);
                }
            }
            else {
                if (entity.isOnGround()) {
                    if (Gdx.input.isKeyPressed(Input.Keys.W)) {
                        setVel(0, JUMP_VERTICAL_IMPULSE);
                    }

                }
                float modifier = (shift ? 2 : 1);

                if (Gdx.input.isKeyPressed(Input.Keys.A)) {
                    setVel(-HORIZONTAL_IMPULSE * modifier, 0);
                }
                if (Gdx.input.isKeyPressed(Input.Keys.D)) {
                    setVel(HORIZONTAL_IMPULSE * modifier, 0);
                }
            }
        }

        if (update) {
            getWorldRender().update();
        }
    }

    private void setVel(float velX, float velY) {
        if (lastEditTick + EDIT_TICK_DELAY <= getWorldRender().getWorld().getTick()) {
            synchronized (BOX2D_LOCK) {
                System.out.println("updating tick (tick " + lastEditTick + ")");
                Body body = getControlled().getBody();
                Vector2 vel = body.getLinearVelocity();
                Vector2 pos = body.getPosition();

                body.applyLinearImpulse(velX == 0 ? vel.x : velX, velY == 0 ? vel.y : velY, pos.x, pos.y, true);
//            body.setLinearVelocity(velX == 0 ? vel.x : velX, velY == 0 ? vel.y : velY);
//            body.setAwake(true);
            }
        }
    }

    @Override
    public boolean keyDown(int keycode) {
        if (Main.inst().getConsole().isVisible()) {
            return false;
        }
        switch (keycode) {
            case NUM_0:
            case NUMPAD_0:
                selected = Material.values()[0];
                break;
            case NUM_1:
            case NUMPAD_1:
                selected = Material.values()[1];
                break;
            case NUM_2:
            case NUMPAD_2:
                selected = Material.values()[2];
                break;
            case NUM_3:
            case NUMPAD_3:
                selected = Material.values()[3];
                break;
            case NUM_4:
            case NUMPAD_4:
                selected = Material.values()[4];
                break;
            case NUM_5:
            case NUMPAD_5:
                selected = Material.values()[5];
                break;
            case NUM_6:
            case NUMPAD_6:
                selected = Material.values()[6];
                break;
            case NUM_7:
            case NUMPAD_7:
                selected = Material.values()[7];
                break;
            case NUM_8:
            case NUMPAD_8:
                selected = Material.values()[8];
                break;
            case NUM_9:
            case NUMPAD_9:
                selected = Material.values()[9];
                break;
            default:
                return false;
        }
        return true;
    }

    @Override
    @Nullable
    public Material getSelected() {
        return selected;
    }

    @Override
    public void setSelected(@Nullable Material selected) {
        this.selected = selected;
    }

    @Override
    public float getBreakBrushSize() {
        return breakBrushSize;
    }

    @Override
    public void setBreakBrushSize(float breakBrushSize) {
        this.breakBrushSize = breakBrushSize;
    }

    @Override
    public float getPlaceBrushSize() {
        return placeBrushSize;
    }

    @Override
    public void setPlaceBrushSize(float placeBrushSize) {
        this.placeBrushSize = placeBrushSize;
    }
}

File: core/src/main/kotlin/no/elg/infiniteBootleg/input/KeyboardControls.kt
package no.elg.infiniteBootleg.input

import com.badlogic.ashley.core.Entity
import com.badlogic.gdx.Gdx
import com.badlogic.gdx.Input.Buttons
import com.badlogic.gdx.Input.Keys
import com.badlogic.gdx.math.Vector2
import no.elg.infiniteBootleg.Settings
import no.elg.infiniteBootleg.main.ClientMain
import no.elg.infiniteBootleg.main.Main
import no.elg.infiniteBootleg.util.breakableBlocks
import no.elg.infiniteBootleg.util.compactLoc
import no.elg.infiniteBootleg.util.dstd
import no.elg.infiniteBootleg.util.placeableBlocks
import no.elg.infiniteBootleg.util.worldToBlock
import no.elg.infiniteBootleg.world.Material
import no.elg.infiniteBootleg.world.ecs.components.InventoryComponent.Companion.inventoryOrNull
import no.elg.infiniteBootleg.world.ecs.components.VelocityComponent.Companion.setVelocity
import no.elg.infiniteBootleg.world.ecs.components.VelocityComponent.Companion.velocity
import no.elg.infiniteBootleg.world.ecs.components.required.PositionComponent.Companion.teleport
import no.elg.infiniteBootleg.world.ecs.components.tags.FlyingTag.Companion.flying
import no.elg.infiniteBootleg.world.ecs.components.transients.Box2DBodyComponent.Companion.box2dBody
import no.elg.infiniteBootleg.world.ecs.components.transients.GroundedComponent.Companion.grounded
import no.elg.infiniteBootleg.world.ecs.components.transients.SelectedInventoryItemComponent.Companion.selectedOrNull
import no.elg.infiniteBootleg.world.ecs.components.transients.WorldComponent.Companion.world
import no.elg.infiniteBootleg.world.ticker.WorldBox2DTicker.Companion.BOX2D_TPS
import no.elg.infiniteBootleg.world.world.ClientWorld
import kotlin.math.abs
import kotlin.math.ceil
import kotlin.math.min

/**
 * Control scheme where the user moves the player around with a keyboard
 *
 * @author Elg
 */
class KeyboardControls(val world: ClientWorld) {

  var brushSize = INITIAL_BRUSH_SIZE
  var interactRadius = INITIAL_INTERACT_RADIUS

  private val tmpVec = Vector2()
  private val tmpVec2 = Vector2()

  private val mouseLocator = MouseLocator()
  private var lastCreateBlockLoc: Long = 0
  private var lastCreateBlockTick: Long = 0

  private fun breakBlocks(entity: Entity, blockX: Int, blockY: Int): Boolean {
    if (canNotInteract(blockX, blockY)) return false
    val world = entity.world
    val breakableBlocks = entity.breakableBlocks(world, blockX, blockY, brushSize, interactRadius).asIterable()
    world.removeBlocks(world.getBlocks(breakableBlocks))
    return true
  }

  private fun placeBlocks(entity: Entity, blockX: Int, blockY: Int): Boolean {
    if (canNotInteract(blockX, blockY)) return false
    val world = entity.world
    val material = (entity.selectedOrNull ?: return false).material
    val inventory = entity.inventoryOrNull ?: return false
    val placeableBlock = entity.placeableBlocks(world, blockX, blockY, brushSize, interactRadius).toSet()
    if (inventory.use(material, placeableBlock.size.toUInt())) {
      material.createBlocks(world, placeableBlock)
    }
    return true
  }

  private fun canNotInteract(blockX: Int, blockY: Int): Boolean {
    val compactLoc = compactLoc(blockX, blockY)
    val tick = world.tick
    if (lastCreateBlockLoc == compactLoc && world.tick - lastCreateBlockTick < BOX2D_TPS / 10f) {
      return true
    }
    lastCreateBlockLoc = compactLoc
    lastCreateBlockTick = tick
    return false
  }

  private fun Entity.fly() {
    fun fly(dx: Float = 0f, dy: Float = 0f) {
      setVel { oldX, oldY -> oldX + dx to oldY + dy }
    }

    when {
      Gdx.input.isKeyPressed(Keys.W) -> fly(dy = FLY_VEL)
      Gdx.input.isKeyPressed(Keys.S) -> fly(dy = -FLY_VEL)
      Gdx.input.isKeyPressed(Keys.A) -> fly(dx = -FLY_VEL)
      Gdx.input.isKeyPressed(Keys.D) -> fly(dx = FLY_VEL)
    }
  }

  private fun Entity.walk() {
    fun moveHorz(dir: Float) {
      val world = world
      world.postBox2dRunnable {
        if (grounded.canMove(dir)) {
          val body = box2dBody

          val currSpeed = body.linearVelocity.x
          val wantedSpeed = dir * if (grounded.onGround) {
            MAX_X_VEL
          } else {
            MAX_X_VEL * (2f / 3f)
          }
          val impulse = body.mass * (wantedSpeed - (dir * min(abs(currSpeed), abs(wantedSpeed))))

          tmpVec.set(impulse, velocity.dy)

          body.applyLinearImpulse(tmpVec, body.worldCenter, true)
        }
      }
    }

    if (Gdx.input.isKeyPressed(Keys.A)) {
      moveHorz(-1f)
    }
    if (Gdx.input.isKeyPressed(Keys.D)) {
      moveHorz(1f)
    }
  }

  private fun Entity.jump() {
    if (this.grounded.onGround && Gdx.input.isKeyPressed(Keys.W)) {
      setVel { oldX, _ -> oldX to JUMP_VERTICAL_VEL }
    }
  }

  fun update(entity: Entity) {
    val entityWorld = entity.world
    if (entityWorld is ClientWorld) {
      mouseLocator.update(entityWorld)
    }
    if (ClientMain.inst().shouldIgnoreWorldInput()) {
      return
    }

    when {
      Gdx.input.isButtonPressed(Buttons.LEFT) -> entity.interpolate(false, this::breakBlocks)
      Gdx.input.isButtonPressed(Buttons.RIGHT) -> entity.interpolate(false, this::placeBlocks)
      Gdx.input.isKeyJustPressed(Keys.Q) -> entity.interpolate(true, this::placeBlocks)
    }

    if (entity.flying) {
      entity.fly()
    } else {
      entity.jump()
      entity.walk()
    }
  }

  private fun Entity.setVel(modify: (oldX: Float, oldY: Float) -> (Pair<Float, Float>)) {
    val body = box2dBody
    val vel = body.linearVelocity
    val (nx, ny) = modify(vel.x, vel.y)
    this.setVelocity(nx, ny)
  }

  fun touchDown(entity: Entity, button: Int) {
    val update =
      when (button) {
        Buttons.LEFT -> entity.interpolate(true, this::breakBlocks)
        Buttons.RIGHT -> entity.interpolate(true, this::placeBlocks)
        else -> false
      }

    if (update) {
      entity.world.render.update()
    }
  }

  fun keyDown(entity: Entity, keycode: Int): Boolean {
    when (keycode) {
      Keys.T -> entity.teleport(mouseLocator.mouseWorldX, mouseLocator.mouseWorldY)
      Keys.Q -> entity.interpolate(true, this::placeBlocks)
    }

    val selectedMaterial = entity.selectedOrNull ?: return true

    val extra = if (Gdx.input.isKeyPressed(Keys.SHIFT_LEFT) || Gdx.input.isKeyPressed(Keys.SHIFT_RIGHT)) 10 else 0
    try {
      when (keycode) {
        Keys.NUM_0, Keys.NUMPAD_0 -> Material.entries[0 + extra]
        Keys.NUM_1, Keys.NUMPAD_1 -> Material.entries[1 + extra]
        Keys.NUM_2, Keys.NUMPAD_2 -> Material.entries[2 + extra]
        Keys.NUM_3, Keys.NUMPAD_3 -> Material.entries[3 + extra]
        Keys.NUM_4, Keys.NUMPAD_4 -> Material.entries[4 + extra]
        Keys.NUM_5, Keys.NUMPAD_5 -> Material.entries[5 + extra]
        Keys.NUM_6, Keys.NUMPAD_6 -> Material.entries[6 + extra]
        Keys.NUM_7, Keys.NUMPAD_7 -> Material.entries[7 + extra]
        Keys.NUM_8, Keys.NUMPAD_8 -> Material.entries[8 + extra]
        Keys.NUM_9, Keys.NUMPAD_9 -> Material.entries[9 + extra]
        else -> null
      }?.let {
        selectedMaterial.material = it
      }
    } catch (_: IndexOutOfBoundsException) {
      // Ignore out of bounds, for materials that don't exist
    }
    return true
  }

  /**
   * @param justPressed Whether to interpolate between placements
   */
  private fun Entity.interpolate(justPressed: Boolean, action: (entity: Entity, blockX: Int, blockY: Int) -> Boolean): Boolean {
    val blockX = mouseLocator.mouseBlockX
    val blockY = mouseLocator.mouseBlockY
    val worldX = mouseLocator.mouseWorldX
    val worldY = mouseLocator.mouseWorldY

    val inSameBlock = mouseLocator.previousMouseBlockX == blockX && mouseLocator.previousMouseBlockY == blockY
    if (justPressed || inSameBlock) {
      return action(this, blockX, blockY)
    }

    val currPos = tmpVec
    val prevPos = tmpVec2

    currPos.x = worldX
    currPos.y = worldY

    prevPos.x = mouseLocator.previousMouseWorldX
    prevPos.y = mouseLocator.previousMouseWorldY

    val distance = currPos.dstd(prevPos)
    // Limit max distance to draw
    val maxDistance = 20.0

    var update = false
    if (distance in 0.0..maxDistance) {
      val iterations = ceil(distance).toInt()

      val logging = Settings.debug && Gdx.graphics.frameId % 100 == 0L && distance > Int.MAX_VALUE
      if (logging) {
        Main.logger().log("---START SMOOTH PLACEMENT---")
        Main.logger().log("(pos) prev: $prevPos, curr $currPos")
        Main.logger().log("(distance) $distance")
        Main.logger().log("Doing $iterations iterations of interpolation")
      }
      for (i in 1 until iterations + 1) {
        val multiplierX = (distance - i) / distance

        // When drawing from (lower right to upper left) and (upper right to lower right) the Y multiplication must be the inverse
        //
        val multiplierY = when {
          prevPos.x < currPos.x && prevPos.y > currPos.y -> 1 - multiplierX
          prevPos.x > currPos.x && prevPos.y < currPos.y -> 1 - multiplierX
          else -> multiplierX
        }

        if (multiplierX.isInfinite() || multiplierX !in 0.0..1.0) {
          // If the multiplier is weird, don't place a block
          continue
        }

        fun halfpoint(i1: Float, i2: Float, multiplier: Double): Float = (min(i1, i2) + abs(i1 - i2) * multiplier).toFloat()

        val pWx = halfpoint(worldX, mouseLocator.previousMouseWorldX, multiplierX)
        val pWy = halfpoint(worldY, mouseLocator.previousMouseWorldY, multiplierY)
        val pBx = worldToBlock(pWx)
        val pBy = worldToBlock(pWy)

        if (logging) {
          Main.logger().log("--inter $i mltX: $multiplierX | mltY: $multiplierY, pBx:$pBx | pBy:$pBy | pWx:$pWx | pWy:$pWy")
        }
        update = update or action(this, pBx, pBy)
      }

      if (logging) {
        Main.logger().log("---END SMOOTH PLACEMENT (update? $update)---")
      }
    }

    return update or action(this, blockX, blockY)
  }

  companion object {

    private const val JUMP_VERTICAL_VEL = 9f
    private const val FLY_VEL = 1f

    const val MAX_X_VEL = 7.5f // ie target velocity
    const val MAX_Y_VEL = 20f

    const val INITIAL_BRUSH_SIZE = 1f
    const val INITIAL_INTERACT_RADIUS = 32f
  }
}
